File: src/main/javabak/events/Registering.java
package redcrafter07.processed.events;

import net.minecraft.core.BlockPos;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.EntityBlock;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.neoforged.bus.api.SubscribeEvent;
import net.neoforged.fml.common.EventBusSubscriber;
import net.neoforged.neoforge.capabilities.Capabilities;
import net.neoforged.neoforge.capabilities.RegisterCapabilitiesEvent;
import net.neoforged.neoforge.client.event.EntityRenderersEvent;
import net.neoforged.neoforge.client.event.RegisterMenuScreensEvent;
import net.neoforged.neoforge.network.event.RegisterPayloadHandlersEvent;
import redcrafter07.processed.ProcessedMod;
import redcrafter07.processed.block.machine_abstractions.BlockSide;
import redcrafter07.processed.block.machine_abstractions.EnergyCapableBlockEntity;
import redcrafter07.processed.block.machine_abstractions.FluidCapableBlockEntity;
import redcrafter07.processed.block.machine_abstractions.ItemCapableBlockEntity;
import redcrafter07.processed.block.tile_entities.FluidTankBlockEntity;
import redcrafter07.processed.block.tile_entities.ModTileEntities;
import redcrafter07.processed.gui.GenericMachineMenuScreen;
import redcrafter07.processed.gui.ModMenuTypes;
import redcrafter07.processed.network.IOChangePacket;
import redcrafter07.processed.network.MultiblockDestroyPacket;
import redcrafter07.processed.network.WrenchModeChangePacket;

import java.util.Arrays;

@EventBusSubscriber(modid = ProcessedMod.ID, bus = EventBusSubscriber.Bus.MOD)
public class Registering {
    private static boolean blockEntityIs(Block block, Class<?> clazz) {
        if (block instanceof EntityBlock entityBlock) {
            try {
                final BlockEntity be = entityBlock.newBlockEntity(BlockPos.ZERO, block.defaultBlockState());
                return clazz.isInstance(be);
            } catch (SecurityException e) {
                return false;
            }
        }
        return false;
    }

    private static boolean isItemCapable(Block block) {
        return blockEntityIs(block, ItemCapableBlockEntity.class);
    }

    private static boolean isFluidCapable(Block block) {
        return blockEntityIs(block, FluidCapableBlockEntity.class);
    }

    private static boolean isEnergyCapable(Block block) {
        return blockEntityIs(block, EnergyCapableBlockEntity.class);
    }

    private static Block[] itemCapable(Block[] block) {
        return Arrays.stream(block).filter(Registering::isItemCapable).toArray(Block[]::new);
    }

    private static Block[] fluidCapable(Block[] block) {
        return Arrays.stream(block).filter(Registering::isFluidCapable).toArray(Block[]::new);
    }

    private static Block[] energyCapable(Block[] block) {
        return Arrays.stream(block).filter(Registering::isEnergyCapable).toArray(Block[]::new);
    }

    @SubscribeEvent
    public static void onRegisterCapabilities(RegisterCapabilitiesEvent event) {
        final var blocks = ModTileEntities.BLOCK_TYPES.getEntries().stream().flatMap(entry -> entry.get().getValidBlocks().stream()).toArray(Block[]::new);

        event.registerBlock(
                Capabilities.ItemHandler.BLOCK,
                (level, pos, state, blockEntity, side) -> {
                    if (blockEntity instanceof ItemCapableBlockEntity itemCapableBlockEntity) {
                        if (side == null) return itemCapableBlockEntity.itemCapabilityForSide(null, state);
                        return itemCapableBlockEntity.itemCapabilityForSide(BlockSide.translateDirection(side, state), state);
                    }
                    return null;
                },
                itemCapable(blocks)
        );

        event.registerBlock(
                Capabilities.FluidHandler.BLOCK,
                (level, pos, state, blockEntity, side) -> {
                    if (blockEntity instanceof FluidCapableBlockEntity itemCapableBlockEntity) {
                        if (side == null) return itemCapableBlockEntity.fluidCapabilityForSide(null, state);
                        return itemCapableBlockEntity.fluidCapabilityForSide(BlockSide.translateDirection(side, state), state);
                    }
                    return null;
                },
                fluidCapable(blocks)
        );
        event.registerBlock(
                Capabilities.EnergyStorage.BLOCK,
                (level, pos, state, blockEntity, side) -> {
                    if (blockEntity instanceof EnergyCapableBlockEntity itemCapableBlockEntity) {
                        if (side == null) return itemCapableBlockEntity.energyCapabilityForSide(null, state);
                        return itemCapableBlockEntity.energyCapabilityForSide(BlockSide.translateDirection(side, state), state);
                    }
                    return null;
                },
                energyCapable(blocks)
        );
    }

    @SubscribeEvent
    public static void registerNetworkHandlers(RegisterPayloadHandlersEvent event) {
        final var register = event.registrar(ProcessedMod.ID);

        register.playToServer(
                WrenchModeChangePacket.TYPE,
                WrenchModeChangePacket.CODEC,
                WrenchModeChangePacket::handleServer
        );

        register.playToServer(
                IOChangePacket.TYPE,
                IOChangePacket.CODEC,
                IOChangePacket::handleServer
        );

        register.playToClient(
                MultiblockDestroyPacket.TYPE,
                MultiblockDestroyPacket.CODEC,
                MultiblockDestroyPacket::handleClient
        );
    }

    @SubscribeEvent
    public static void registerMenuScreens(RegisterMenuScreensEvent event) {
        event.register(
                ModMenuTypes.POWERED_FURNACE_MENU.get(),
                GenericMachineMenuScreen::new
        );
    }

    @SubscribeEvent
    public static void registerRenderers(EntityRenderersEvent.RegisterRenderers event) {
        event.registerBlockEntityRenderer(ModTileEntities.FLUID_TANK.get(), FluidTankBlockEntity.FluidTankEntityRenderer::new);
    }
}

File: src/main/kotlin/redcrafter07/processed/events/Registering.kt
package redcrafter07.processed.events

import net.minecraft.core.BlockPos
import net.minecraft.world.level.block.Block
import net.minecraft.world.level.block.EntityBlock
import net.neoforged.bus.api.SubscribeEvent
import net.neoforged.fml.ModList
import net.neoforged.fml.common.EventBusSubscriber
import net.neoforged.fml.event.lifecycle.FMLLoadCompleteEvent
import net.neoforged.neoforge.capabilities.Capabilities
import net.neoforged.neoforge.capabilities.RegisterCapabilitiesEvent
import net.neoforged.neoforge.client.event.RegisterParticleProvidersEvent
import net.neoforged.neoforge.event.AddPackFindersEvent
import net.neoforged.neoforge.fluids.capability.wrappers.FluidBucketWrapper
import net.neoforged.neoforge.network.event.RegisterPayloadHandlersEvent
import net.neoforged.neoforge.registries.DataPackRegistryEvent
import net.neoforged.neoforge.registries.NewRegistryEvent
import redcrafter07.processed.ProcessedMod
import redcrafter07.processed.ProcessedPower
import redcrafter07.processed.block.machine_abstractions.BlockSide
import redcrafter07.processed.block.machine_abstractions.EnergyCapableBlockEntity
import redcrafter07.processed.block.machine_abstractions.FluidCapableBlockEntity
import redcrafter07.processed.block.machine_abstractions.ItemCapableBlockEntity
import redcrafter07.processed.block.tile_entities.ModTileEntities
import redcrafter07.processed.covers.Cover
import redcrafter07.processed.dynpack.DynPackSource
import redcrafter07.processed.fluid.ModFluids
import redcrafter07.processed.integration.theoneprobe.TheOneProbeIntegration
import redcrafter07.processed.miner.MinerData
import redcrafter07.processed.miner.Planetoid
import redcrafter07.processed.network.*
import redcrafter07.processed.particles.FireParticle.FireParticleProvider
import redcrafter07.processed.particles.ModParticles
import redcrafter07.processed.particles.SmokeParticle.SmokeParticleProvider
import java.util.*

@EventBusSubscriber(modid = ProcessedMod.ID, bus = EventBusSubscriber.Bus.MOD)
object Registering {
    /** Checks if the block entity associated with this block implements the clazz. */
    private fun blockEntityIs(block: Block?, clazz: Class<*>): Boolean {
        if (block is EntityBlock) {
            try {
                val be = block.newBlockEntity(BlockPos.ZERO, block.defaultBlockState())
                return clazz.isInstance(be)
            } catch (_: SecurityException) {
                return false
            }
        }
        return false
    }

    /** Checks if the block entity associated with this block implements ItemCapableBlockEntity */
    private fun isItemCapable(block: Block): Boolean {
        return blockEntityIs(block, ItemCapableBlockEntity::class.java)
    }

    /** Checks if the block entity associated with this block implements FluidCapableBlockEntity */
    private fun isFluidCapable(block: Block): Boolean {
        return blockEntityIs(block, FluidCapableBlockEntity::class.java)
    }

    /** Checks if the block entity associated with this block implements EnergyCapableBlockEntity */
    private fun isEnergyCapable(block: Block): Boolean {
        return blockEntityIs(block, EnergyCapableBlockEntity::class.java)
    }

    /** Filters out any blocks that don't implement ItemCapableBlockEntity */
    private fun itemCapable(block: Array<Block>): Array<Block> =
        Arrays.stream(block).filter(Registering::isItemCapable).toArray { arrayOfNulls(it) }

    /** Filters out any blocks that don't implement FluidCapableBlockEntity */
    private fun fluidCapable(block: Array<Block>): Array<Block> =
        Arrays.stream(block).filter(Registering::isFluidCapable).toArray { arrayOfNulls(it) }

    /** Filters out any blocks that don't implement EnergyCapableBlockEntity */
    private fun energyCapable(block: Array<Block>): Array<Block> =
        Arrays.stream(block).filter(Registering::isEnergyCapable).toArray { arrayOfNulls(it) }

    @SubscribeEvent
    fun onRegisterCapabilities(event: RegisterCapabilitiesEvent) {
        // Gets all processed blocks
        val blocks = ModTileEntities.BLOCK_TYPES.entries.stream().flatMap { it.get().validBlocks.stream() }.toList()
            .toTypedArray()

        // registers all blocks that have an item capability as having an item capability
        event.registerBlock(
            Capabilities.ItemHandler.BLOCK,
            { _, _, state, blockEntity, side ->
                if (blockEntity is ItemCapableBlockEntity) {
                    if (side == null) blockEntity.itemCapabilityForSide(null, state)
                    else blockEntity.itemCapabilityForSide(BlockSide.translateDirection(side, state), state)
                } else null
            },
            *itemCapable(blocks),
        )
        // registers all blocks that have a processed energy capability as having an item capability
        event.registerBlock(
            ProcessedPower.BLOCK,
            { _, _, state, blockEntity, side ->
                if (blockEntity is EnergyCapableBlockEntity) {
                    if (side == null) blockEntity.energyCapabilityForSide(null, state)
                    else blockEntity.energyCapabilityForSide(BlockSide.translateDirection(side, state), state)
                } else null
            },
            *energyCapable(blocks),
        )
        // registers all blocks that have a fluid as having an item capability
        event.registerBlock(
            Capabilities.FluidHandler.BLOCK,
            { _, _, state, blockEntity, side ->
                if (blockEntity is FluidCapableBlockEntity) {
                    if (side == null) blockEntity.fluidCapabilityForSide(null, state)
                    else blockEntity.fluidCapabilityForSide(BlockSide.translateDirection(side, state), state)
                } else null
            },
            *fluidCapable(blocks),
        )

        event.registerItem(
            Capabilities.FluidHandler.ITEM,
            { item, _ -> FluidBucketWrapper(item) },
            *ModFluids.REGISTERED_FLUIDS.map { it.bucket }.toList().toTypedArray()
        )
    }

    @SubscribeEvent
    // registers all packets
    fun registerNetworkHandlers(event: RegisterPayloadHandlersEvent) {
        val registrar = event.registrar(ProcessedMod.ID)

        registrar.playBidirectional(MenuRPCPacket.TYPE, MenuRPCPacket.CODEC, MenuRPCPacket::handle)
        registrar.playBidirectional(RPCPacket.TYPE, RPCPacket.CODEC, RPCPacket::handle)
    }


    @SubscribeEvent
    fun registerPackSources(event: AddPackFindersEvent) = event.addRepositorySource(DynPackSource)

    @SubscribeEvent
    fun loadComplete(e: FMLLoadCompleteEvent) {
        e.enqueueWork { if (ModList.get().isLoaded("theoneprobe")) TheOneProbeIntegration.init() }
    }

    @SubscribeEvent
    fun registerDatapackRegistries(e: DataPackRegistryEvent.NewRegistry) {
        e.dataPackRegistry(
            Planetoid.REGISTRY_KEY,
            Planetoid.CODEC,
            Planetoid.CODEC,
        )
        e.dataPackRegistry(
            MinerData.Fuel.REGISTRY_KEY,
            MinerData.Fuel.CODEC,
            MinerData.Fuel.CODEC,
        )
    }

    @SubscribeEvent
    fun registerParticleProviders(e: RegisterParticleProvidersEvent) {
        e.registerSpriteSet(ModParticles.SMOKE.get(), ::SmokeParticleProvider)
        e.registerSpriteSet(ModParticles.FIRE.get(), ::FireParticleProvider)
    }

    @SubscribeEvent
    fun registerRegistries(e: NewRegistryEvent) {
        e.register(Cover.REGISTRY)
    }
}